be/src/exec/sorting/merge_path.h

```diff
@@ -0,0 +1,435 @@
+     * @param start_di Start index of the whole merge result, NOT only the above dest.
+     * @param length The step for this parallelism can forward along merge path.
+     */
+void _do_merge_along_merge_path(const SortDescs& descs, const InputSegment& left, size_t& li, const InputSegment& right,
+                                size_t& ri, OutputSegment& dest, size_t start_di, const size_t length);
+} // namespace detail
+
+class MergePathCascadeMerger;
+
+/**
+ * LeafNode will use this provider to get data.
+ *
+ * @param only_check_if_has_data If true, only check if it has more data for further fetching
+ * @param chunk Output chunk if any
+ * @param eos Set to true if this provider reaches eos
+ * @return: Return true if it has new data
+ */
+using MergePathChunkProvider = std::function<bool(bool only_check_if_has_data, ChunkPtr* chunk, bool* eos)>;
+
+namespace detail {
+
+// A merge tree is composed by a group of node, node can be either LeafNode or MergeNode
+// LeafNode gets data from the specific providers, and has no child
+// MergeNode has two children, and combine two streams of data into one.
+class Node {
+public:
+    Node(MergePathCascadeMerger* merger) : _merger(merger) {}
+    virtual ~Node() = default;
+    virtual void process_input(int32_t parallel_idx) = 0;
+    virtual void process_input_done(){};
+    virtual bool is_leaf() { return false; }
+    // Means the previous node reaches eos, no more data will flow to current node,
+    // but there still may be some data in the input buffer to be consumed.
+    virtual bool dependency_finished() = 0;
+    // Means based on dependency_finished state, all the data in the input buffer has been consumed.
+    virtual bool input_finished() = 0;
+    // Return true if current node is active but cannot provide more data at this moment.
+    virtual bool is_pending() { return false; }
+    size_t degree_of_parallelism() { return _global_2_local_parallel_idx.size(); }
+    void bind_parallel_idxs(std::unordered_map<int32_t, int32_t>&& global_2_local_parallel_idx) {
+        _global_2_local_parallel_idx = std::move(global_2_local_parallel_idx);
+    }
+    bool output_empty() { return _output_segments.empty(); }
+    bool eos() { return dependency_finished() && input_finished() && output_empty(); }
+    std::vector<OutputSegmentPtr>&& output_segments() { return std::move(_output_segments); }
+
+protected:
+    MergePathCascadeMerger* _merger;
+
+    // There may be multiply workers work on the same Node. So wee need to know
+    // the local parallel_idx which the global parallel_idx corresponds to.
+    // For example, there are 4 workers with id 0-3, and there are 2 Node(A and B).
+    // Each Node should have two workers, assuming Node A have worker_0 and worker_1,
+    // Node B have worker_2 and worker_3. So Node A need to maintain the mapping of {0->0, 1->0},
+    // and Node B need to maintain the mapping of {2->0, 3->0}, given that the local parallel_idx
```

Comment:
{0->0, 1->1}, {2->0, 3->1}?

Reply:
yes, mapping global parallel_idx to local_parallel_idx (each node has its owe view)

Reply:
but your comment is wrong.

Reply:
I got it.